open-models/serving/cloud_run_ollama_gemma2_rag_qa.ipynb

```diff
@@ -0,0 +1,1144 @@
+        "For more information about your identity tokens expiring and how to refresh your tokens, see the next section below \"Appendix: Handling Identity Token Expiration\".\n",
+        "\n",
+        "To clean up the resources you created in this section, see the section at the bottom \"Cleaning up\"."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "A5VPH4gP3igf"
+      },
+      "source": [
+        "## Appendix: Handling Identity Token Expiration\n",
+        "\n",
+        "When deploying a Generative AI application Google Cloud Run, you'll often need to authenticate your requests using Identity Tokens.\n",
+        "\n",
+        "These tokens will expire hourly, requiring a mechanism for automatic refresh to ensure uninterrupted operation.\n",
+        "\n",
+        "The following helper classes provide an example of how to deal with  token refresh. It leverages the `google.auth` library to handle the authentication process and automatically refresh the token when necessary.\n",
+        "\n",
+        "\n",
+        "See the following resources for more information on authentication:\n",
+        "* [Identity Token Overview](https://cloud.google.com/docs/authentication/token-types#id)\n",
+        "* [Google Cloud Run Authentication Documentation](https://cloud.google.com/run/docs/authenticating/overview)\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "TJlxAZKz-tjY"
+      },
+      "outputs": [],
+      "source": [
+        "import time\n",
+        "\n",
+        "import google.auth\n",
+        "from google.auth.exceptions import DefaultCredentialsError\n",
+        "import google.auth.transport.requests\n",
+        "import google.oauth2.id_token\n",
+        "from pydantic.v1 import Extra\n",
+        "\n",
+        "\n",
+        "class TokenManager:\n",
+        "    def __init__(\n",
+        "        self, url=None, token_lifetime=3600\n",
+        "    ):  # Default token lifetime of 1 hour\n",
+        "        self.token = None\n",
+        "        self.expiry_time = 0\n",
+        "        self.token_lifetime = token_lifetime\n",
+        "        self.url = url\n",
+        "        self.creds, _ = google.auth.default()\n",
+        "\n",
+        "    def get_token(self):\n",
+        "        if time.time() >= self.expiry_time:\n",
+        "            self.refresh_token(url=self.url)\n",
+        "        return self.token\n",
+        "\n",
+        "    def refresh_token(self, url):\n",
+        "        \"\"\"\n",
+        "        Retrieves an ID token, attempting to use default credentials first,\n",
+        "        and falling back to fetching a service-to-service new token if necessary.\n",
+        "        See more on Cloud Run authentication at this link:\n",
+        "         https://cloud.google.com/run/docs/authenticating/service-to-service\n",
+        "        Args:\n",
```

Comment:
Please add types for parameters/return value.

Reply:
done, thanks!